service/src/tap_manager.rs

```diff
@@ -0,0 +1,163 @@
+// Copyright 2023-, GraphOps and Semiotic Labs.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::sync::Arc;
+
+use alloy_sol_types::Eip712Domain;
+use log::error;
+use sqlx::{types::BigDecimal, PgPool};
+use tap_core::tap_manager::SignedReceipt;
+
+use crate::{allocation_monitor, escrow_monitor, query_processor::QueryError};
+
+#[derive(Clone, Debug)]
+pub struct TapManager {
+    allocation_monitor: allocation_monitor::AllocationMonitor,
+    escrow_monitor: escrow_monitor::EscrowMonitor,
+    pgpool: PgPool,
+    domain_separator: Arc<Eip712Domain>,
+}
+
+impl TapManager {
+    pub fn new(
+        pgpool: PgPool,
+        allocation_monitor: allocation_monitor::AllocationMonitor,
+        escrow_monitor: escrow_monitor::EscrowMonitor,
+        domain_separator: Eip712Domain,
+    ) -> Self {
+        Self {
+            allocation_monitor,
+            escrow_monitor,
+            pgpool,
+            domain_separator: Arc::new(domain_separator),
+        }
+    }
+
+    /// Checks that the receipt refers to eligible allocation ID and TAP sender.
+    ///
+    /// If the receipt is valid, it is stored in the database.
+    ///
+    /// The rest of the TAP receipt checks are expected to be performed out-of-band by the receipt aggregate requester
+    /// service.
+    pub async fn verify_and_store_receipt(&self, receipt: SignedReceipt) -> Result<(), QueryError> {
```

Comment:
Well, this is technically a query error that we want to communicate back to the consumer, because here an error can happen only if the sender's receipt is invalid.

Reply:
Ah okay, understood